# Fix downloaded dataset

### Redownload missing data

In [14]:
import os
import csv
import time
import requests

In [2]:
fdir = '../../../data/databases/Sabio-RK/'
fname = 'dataset_download.tsv'
f = open(fdir+fname, 'r')
lines = f.readlines()
f.close()

Certain database entries can be downloaded incompletely with the last line mostly missing (without `\n`) for some ungodly reason. Manually fixing these missing entries in the corresponding text files by redownloading the corresponding entries.

In [3]:
for line in lines:
    data = line.strip().split('\t')
    if len(data) != 17:
        print('EntryID:', data[0])
        print(line)

EntryID: 57731
57731	7325957	Paenarthrobacter ureafaciens	GM3;H2O	wildtype	exo-alpha-s58	15195997	Escherichia coli	Peptide;H2O	mutant SPase I  R127A	signal peptidase I	P00803	3.4.21.89	E	concentration	unknown compound	1.69E-6		-	M	7.7	37.0

EntryID: 58094
58094	23758290	Bacillus anthracis	H2O;Bacillithiol	wildtype His6-tagged	In linear amides	Q81ST8	3.5.1	S	concentration	Bacillithiol	2.0E-4	0.01	-	M	7.4	5810	16036918	Escherichia coli	Chorismate	wildtype	chorismate mutase	P0A9J8	5.4.99.5	A	concentration	Chorismate	7.0E-6	0.0023	-	M	7.5	30.0

EntryID: 58472
58472	8706715	Astragalus bisulcatus	S-Adenosyl-L-methionine;L-Selenocysteine	wildtype	selenocysteine Se-methyltransferase	P56707	2.1.1.280	Km_A	Km	S-A5848	15642737	Streptococcus agalactiae	L-gamma-Glutamyl-L-alpha-aminobutyrate;Glycine;ATP	wildtype	glutathione synthase	Q8DXM9	6.3.2.3	A	concentration	Glycine	0.025		-	M	8.4	37.0

EntryID: 60650
60650	16415881	Homo sapiens	S-Adenosyl-L-methionine;TAF10-K189	wildtype	histone-lysine N-meth

In [15]:
QUERY_URL = 'http://sabiork.h-its.org/sabioRestWebServices/kineticlawsExportTsv'
#query = {'fields[]':['EntryID', 'Substrate', 'EnzymeType', 'Enzymename', 'PubMedID', 'Organism', 'UniprotID', 'ECNumber', 'Parameter', 'pH', 'Temperature']}
query = {'fields[]':['EntryID', 'PubMedID', 'Organism', 'Substrate', 'EnzymeType', 'Enzymename', 'UniprotID', 'ECNumber', 'Parameter', 'pH', 'Temperature']}

In [16]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:57731"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

2.1117329597473145
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
57731	7325957	Paenarthrobacter ureafaciens	H2O;GM3	wildtype	exo-alpha-sialidase	Q5W7Q2	3.2.1.18	Vmax	Vmax		1.95E-4		-	mol*s^(-1)*g^(-1)	4.8	37.0
57731	7325957	Paenarthrobacter ureafaciens	H2O;GM3	wildtype	exo-alpha-sialidase	Q5W7Q2	3.2.1.18	Km	Km	GM3	8.0E-5		-	M	4.8	37.0



In [17]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:58094"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

1.9569458961486816
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
58094	23758290	Bacillus anthracis	H2O;Bacillithiol	wildtype His6-tagged	In linear amides	Q81ST8	3.5.1	E	concentration	unknown compound	5.0E-4	0.1	-	mg/ml	7.4	37.0
58094	23758290	Bacillus anthracis	H2O;Bacillithiol	wildtype His6-tagged	In linear amides	Q81ST8	3.5.1	kcat_Km	kcat/Km	Bacillithiol	1.4		1.8E-1	M^(-1)*s^(-1)	7.4	37.0
58094	23758290	Bacillus anthracis	H2O;Bacillithiol	wildtype His6-tagged	In linear amides	Q81ST8	3.5.1	S	concentration	Bacillithiol	2.0E-4	0.01	-	M	7.4	37.0
58094	23758290	Bacillus anthracis	H2O;Bacillithiol	wildtype His6-tagged	In linear amides	Q81ST8	3.5.1	Km	Km	Bacillithiol	4.8E-4		5E-5	M	7.4	37.0
58094	23758290	Bacillus anthracis	H2O;Bacillithiol	wildtype His6-tagged	In linear amides	Q81ST8	3.5.1	kcat	kcat		7.0E-

In [18]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:58472"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

1.9602129459381104
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
58472	8706715	Astragalus bisulcatus	L-Selenocysteine;S-Adenosyl-L-methionine	wildtype	selenocysteine Se-methyltransferase	P56707	2.1.1.280	Km_B	Km	L-Selenocysteine	7.0E-4		-	M	6.0	30.0
58472	8706715	Astragalus bisulcatus	L-Selenocysteine;S-Adenosyl-L-methionine	wildtype	selenocysteine Se-methyltransferase	P56707	2.1.1.280	Km_A	Km	S-Adenosyl-L-methionine	1.5E-5		-	M	6.0	30.0



In [19]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:60650"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

2.02406907081604
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
60650	16415881	Homo sapiens	S-Adenosyl-L-methionine;TAF10-K189	wildtype	histone-lysine N-methyltransferase	Q12962 Q8WTS6	2.1.1.43	kcat_Km	kcat/Km	TAF10-K189	5.666667E8		5E7	M^(-1)*s^(-1)	7.0	30.0
60650	16415881	Homo sapiens	S-Adenosyl-L-methionine;TAF10-K189	wildtype	histone-lysine N-methyltransferase	Q12962 Q8WTS6	2.1.1.43	E	concentration	unknown compound			-	-	7.0	30.0
60650	16415881	Homo sapiens	S-Adenosyl-L-methionine;TAF10-K189	wildtype	histone-lysine N-methyltransferase	Q12962 Q8WTS6	2.1.1.43	B	concentration	S-Adenosyl-L-methionine	5.0E-5		-	M	7.0	30.0
60650	16415881	Homo sapiens	S-Adenosyl-L-methionine;TAF10-K189	wildtype	histone-lysine N-methyltransferase	Q12962 Q8WTS6	2.1.1.43	kcat	kcat		0.006833333		1.666667E-4	s^(-1)	7.0	30.0
606

In [20]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:61738"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

1.8870289325714111
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
61738	10995231	Mycobacterium tuberculosis	Guanosine 3'-diphosphate 5'-triphosphate;H2O	wildtype	guanosine-3',5'-bis(diphosphate) 3'-diphosphatase	P9WHG9	3.1.7.2	kcat_Km	kcat/Km	Guanosine 3'-diphosphate 5'-triphosphate	4280.0		-	M^(-1)*s^(-1)	8.0	30.0
61738	10995231	Mycobacterium tuberculosis	Guanosine 3'-diphosphate 5'-triphosphate;H2O	wildtype	guanosine-3',5'-bis(diphosphate) 3'-diphosphatase	P9WHG9	3.1.7.2	S	concentration	Guanosine 3'-diphosphate 5'-triphosphate			-	-	8.0	30.0
61738	10995231	Mycobacterium tuberculosis	Guanosine 3'-diphosphate 5'-triphosphate;H2O	wildtype	guanosine-3',5'-bis(diphosphate) 3'-diphosphatase	P9WHG9	3.1.7.2	kcat	kcat		2.61		4E-1	s^(-1)	8.0	30.0
61738	10995231	Mycobacterium tuberculosis	Guanosine 3'-diphosphat

In [21]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:65338"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

2.0431010723114014
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
65338	10091606	Helicobacter pylori	Oxaloacetate;Acetyl-CoA;H2O	wildtype	citrate synthase (unknown stereospecificity)	P56062	2.3.3.16	E	concentration	unknown compound	0.489		-	mg/ml	7.4	37.0
65338	10091606	Helicobacter pylori	Oxaloacetate;Acetyl-CoA;H2O	wildtype	citrate synthase (unknown stereospecificity)	P56062	2.3.3.16	Km_B	Km	Oxaloacetate	3.47E-4		4.3E-5	M	7.4	37.0
65338	10091606	Helicobacter pylori	Oxaloacetate;Acetyl-CoA;H2O	wildtype	citrate synthase (unknown stereospecificity)	P56062	2.3.3.16	B	concentration	Oxaloacetate	0.005		-	M	7.4	37.0
65338	10091606	Helicobacter pylori	Oxaloacetate;Acetyl-CoA;H2O	wildtype	citrate synthase (unknown stereospecificity)	P56062	2.3.3.16	Vmax	Vmax		2.816667E-6		1.833333E-7	mol*s^(-1)*g^(-1)	7.4	37.0

In [22]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:68232"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

1.954827070236206
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
68232	19645480	SARS coronavirus	Arg-Leu-Arg-Gly-Gly-AMC;H2O	wildtype PLpro (aa 1541-1855)		P0C6U8 P0C6X7	3.4.22.n14	vi	rate const.				-	-	7.5	-
68232	19645480	SARS coronavirus	Arg-Leu-Arg-Gly-Gly-AMC;H2O	wildtype PLpro (aa 1541-1855)		P0C6U8 P0C6X7	3.4.22.n14	I	concentration	5-Amino-2-methyl-N-[(1R)-1-(1-naphthyl)ethyl]benzamide	0.0	2.0E-4	-	M	7.5	-
68232	19645480	SARS coronavirus	Arg-Leu-Arg-Gly-Gly-AMC;H2O	wildtype PLpro (aa 1541-1855)		P0C6U8 P0C6X7	3.4.22.n14	IC50	IC50	5-Amino-2-methyl-N-[(1R)-1-(1-naphthyl)ethyl]benzamide	5.6E-7		3E-8	M	7.5	-
68232	19645480	SARS coronavirus	Arg-Leu-Arg-Gly-Gly-AMC;H2O	wildtype PLpro (aa 1541-1855)		P0C6U8 P0C6X7	3.4.22.n14	E	concentration	unknown compound	1.25E-7		-	M	7.5	-
68232	19645480	SARS coronav

In [23]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:71116"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

2.1076080799102783
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
71116	15855156	Micrococcus luteus	ATP;NADH	wildtype	NADH kinase	Q8VUL9	2.7.1.86	E	concentration	unknown compound	0.004		-	mg/ml	7.0	37.0
71116	15855156	Micrococcus luteus	ATP;NADH	wildtype	NADH kinase	Q8VUL9	2.7.1.86	A	concentration	NADH	0.0025		-	M	7.0	37.0
71116	15855156	Micrococcus luteus	ATP;NADH	wildtype	NADH kinase	Q8VUL9	2.7.1.86	Vmax	Vmax		1.441667E-4		4.85E-6	mol*s^(-1)*g^(-1)	7.0	37.0
71116	15855156	Micrococcus luteus	ATP;NADH	wildtype	NADH kinase	Q8VUL9	2.7.1.86	Km	Km	ATP	3.2E-4		6E-6	M	7.0	37.0
71116	15855156	Micrococcus luteus	ATP;NADH	wildtype	NADH kinase	Q8VUL9	2.7.1.86	S	concentration	ATP			-	-	7.0	37.0



In [24]:
# Extra code to rerun certain chunks/entries if needed 
start = time.time()
query_string = "EntryID:75485"
query['q'] = query_string
request = requests.get(QUERY_URL, params = query)
request.raise_for_status()
end = time.time()
print(end - start)
print(request.text)

2.0773558616638184
EntryID	PubMedID	Organism	Substrate	EnzymeType	Enzymename	UniprotID	ECNumber	parameter.name	parameter.type	parameter.associatedSpecies	parameter.startValue	parameter.endValue	parameter.standardDeviation	parameter.unit	pH	Temperature
75485	29807982	Papaver somniferum	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	thebaine synthase	A0A2U9GGW3	4.2.99.24	S	concentration	7-O-Acetylsalutaridinol	0.0	1.0E-4	-	M	7.0	30.0
75485	29807982	Papaver somniferum	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	thebaine synthase	A0A2U9GGW3	4.2.99.24	n	Hill coefficient	7-O-Acetylsalutaridinol	2.316		3.52E-1	-	7.0	30.0
75485	29807982	Papaver somniferum	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	thebaine synthase	A0A2U9GGW3	4.2.99.24	Vmax	Vmax		5.36E-5		3.05E-6	mol*s^(-1)*g^(-1)	7.0	30.0
75485	29807982	Papaver somniferum	7-O-Acetylsalutaridinol	wildtype THS2 His6-tagged	thebaine synthase	A0A2U9GGW3	4.2.99.24	S_half	S_half	7-O-Acetylsalutaridinol	2.197E-5		2.036E-6	M	7.0	30.0


### Resolve inconsistent formatting

In [25]:
fdir = '../../../data/databases/Sabio-RK/'
fname = 'dataset_download.tsv'
with open(fdir+fname, 'r') as f:
    data = f.read()

In [26]:
# remove unnecessary dashes and spaces so the corresponding entries are treated as missing values
data = data.replace("\t-", "\t")
data = data.replace("\t ", "\t")
data = data.replace("\t-\n", "\t\n")

In [27]:
fname = 'Sabio-RK_dataset.tsv'
with open(fdir+fname, 'w') as f:
    f.write(data)

# Extract and clean up $k_{cat}$ entries

### Load data

Load the dataset as pandas dataframe and inspect the specifics of each parameter and their associated values

In [28]:
import pandas as pd
import numpy as np
pd.options.mode.chained_assignment = None

In [29]:
fdir = '../../../data/databases/Sabio-RK/'
fname = 'Sabio-RK_dataset.tsv'
df = pd.read_csv(fdir+fname, sep='\t')

In [30]:
len(np.unique(df['EntryID']))

74395

In [31]:
df

EntryID    PubMedID              Organism  \
0             1  12962477.0       Bacillus lentus   
1             1  12962477.0       Bacillus lentus   
2             1  12962477.0       Bacillus lentus   
3             1  12962477.0       Bacillus lentus   
4            10  15311930.0    Pseudomonas putida   
...         ...         ...                   ...   
334081     9997  15660220.0  Phlebiopsis gigantea   
334082     9998  15660220.0  Phlebiopsis gigantea   
334083     9998  15660220.0  Phlebiopsis gigantea   
334084     9998  15660220.0  Phlebiopsis gigantea   
334085     9999  15660220.0  Phlebiopsis gigantea   

                                   Substrate  \
0                                H2O;Peptide   
1                                H2O;Peptide   
2                                H2O;Peptide   
3                                H2O;Peptide   
4       Riboflavin-5-phosphate;(S)-Mandelate   
...                                      ...   
334081                  O2;2-Deoxy-D-glucose   
334082      O2;beta-Methyl-D-glucopyranoside   
334083      O2;beta-Methyl-D-glucopyranoside   
334084      O2;beta-Methyl-D-glucopyranoside   
334085      beta-Methyl-D-glucopyranoside;O2   

                                           EnzymeType  \
0       wildtype variant DSAI (N76D/N87S/S103A/V104I)   
1       wildtype variant DSAI (N76D/N87S/S103A/V104I)   
2       wildtype variant DSAI (N76D/N87S/S103A/V104I)   
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)   
4                                     mutant MDH G81S   
...                                               ...   
334081              mutant P2OxB2H (K312E-E540K-His6)   
334082                                  wildtype P2Ox   
334083                                  wildtype P2Ox   
334084                                  wildtype P2Ox   
334085                          mutant P2OxB1 (E540K)   

                         Enzymename UniprotID   ECNumber parameter.name  \
0                        subtilisin       NaN  3.4.21.62             Km   
1                        subtilisin       NaN  3.4.21.62             Ki   
2                        subtilisin       NaN  3.4.21.62        kcat/Km   
3                        subtilisin       NaN  3.4.21.62           kcat   
4       (S)-mandelate dehydrogenase    P20932  1.1.99.31           kcat   
...                             ...       ...        ...            ...   
334081             pyranose oxidase    Q6UG02   1.1.3.10             Km   
334082             pyranose oxidase    Q6UG02   1.1.3.10           Vmax   
334083             pyranose oxidase    Q6UG02   1.1.3.10        kcat_Km   
334084             pyranose oxidase    Q6UG02   1.1.3.10             Km   
334085             pyranose oxidase    Q6UG02   1.1.3.10             Km   

       parameter.type                        parameter.associatedSpecies  \
0                  Km                                            Peptide   
1                  Ki  N-(Carbobenzoxy)-Leu-Leu-Phe-trifluoromethylke...   
2             kcat/Km                                            Peptide   
3                kcat                                                NaN   
4                kcat                                                NaN   
...               ...                                                ...   
334081             Km                                  2-Deoxy-D-glucose   
334082           Vmax                                                NaN   
334083        kcat/Km                      beta-Methyl-D-glucopyranoside   
334084             Km                      beta-Methyl-D-glucopyranoside   
334085             Km                      beta-Methyl-D-glucopyranoside   

        parameter.startValue  parameter.endValue  parameter.standardDeviation  \
0               1.900000e-03                 NaN                          NaN   
1               1.800000e-08                 NaN                          NaN   
2               1.200000e+05                 NaN    

### Extract $k_{cat}$ values

In [32]:
mask = df['parameter.type'] == 'kcat'
df = df.loc[mask]
n0 = len(df)
n0

29588

Remove missing $k_{cat}$ values

In [33]:
mask = ~df['parameter.startValue'].isnull()
print("Number of entries with no assigned value:", n0 - np.sum(mask))
df = df.loc[mask]

Number of entries with no assigned value: 1368


In [34]:
print("Remaining number of entries:", len(df))

Remaining number of entries: 28220


Remove unnecessary `parameter.name`, `parameter.type` and `parameter.associatedSpecies` columns

In [35]:
df = df.drop(['parameter.name', 'parameter.type', 'parameter.associatedSpecies'], axis=1)

In [36]:
df

EntryID    PubMedID                  Organism  \
3             1  12962477.0           Bacillus lentus   
4            10  15311930.0        Pseudomonas putida   
7           100  15740745.0          Escherichia coli   
13          101  15740745.0          Escherichia coli   
16          102  15740745.0          Escherichia coli   
...         ...         ...                       ...   
333508     9874  12912903.0  Arthrobacter globiformis   
333511     9875  12912903.0  Arthrobacter globiformis   
333512     9876  12912903.0  Arthrobacter globiformis   
333976     9970  11141073.0      Mesocricetus auratus   
333983     9971  11141073.0      Mesocricetus auratus   

                                        Substrate  \
3                                     H2O;Peptide   
4            Riboflavin-5-phosphate;(S)-Mandelate   
7                          Dihydrofolate;NADPH;H+   
13                         NADPH;Dihydrofolate;H+   
16                         H+;Dihydrofolate;NADPH   
...                                           ...   
333508                 N,N-Dimethylglycine;O2;H2O   
333511                 H2O;N,N-Dimethylglycine;O2   
333512                 H2O;O2;N,N-Dimethylglycine   
333976  NADH;5alpha-Androstane-17beta-ol-3-one;H+   
333983  5alpha-Androstane-3alpha,17beta-diol;NAD+   

                                           EnzymeType  \
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)   
4                                     mutant MDH G81S   
7                                       wildtype DHFR   
13                                   mutant DHFR W30A   
16                                   mutant DHFR W30E   
...                                               ...   
333508                                       wildtype   
333511                                   mutant H225Q   
333512                                   mutant Y259F   
333976                                  wildtype P26h   
333983                                  wildtype P26h   

                         Enzymename UniprotID   ECNumber  \
3                        subtilisin       NaN  3.4.21.62   
4       (S)-mandelate dehydrogenase    P20932  1.1.99.31   
7           dihydrofolate reductase    P0ABQ4    1.5.1.3   
13          dihydrofolate reductase    P0ABQ4    1.5.1.3   
16          dihydrofolate reductase    P0ABQ4    1.5.1.3   
...                             ...       ...        ...   
333508      dimethylglycine oxidase       NaN   1.5.3.10   
333511      dimethylglycine oxidase       NaN   1.5.3.10   
333512      dimethylglycine oxidase       NaN   1.5.3.10   
333976   carbonyl reductase (NADPH)    Q9Z245  1.1.1.184   
333983   carbonyl reductase (NADPH)    Q9Z245  1.1.1.184   

        parameter.startValue  parameter.endValue  parameter.standardDeviation  \
3                   220.0000                 NaN                          NaN   
4                     2.8000                 NaN                          0.1   
7                     6.1800                 NaN                          NaN   
13                   14.5000                 NaN                          NaN   
16                   11.5800                 NaN                          NaN   
...                      ...                 ...                          ...   
333508               10.6000                 NaN                          NaN   
333511                4.3500                 NaN                          NaN   
333512                0.0047                 NaN                          NaN   
333976                0.8500                 NaN                          NaN   
333983                1.2800                 NaN                          NaN   

       parameter.unit   pH  Temperature  
3              s^(-1)  7.5         25.0  
4              s^(-1)  7.5         20.0  
7              s^(-1)  7.0         15.0  
13             s^(-1)  7.0         15.0  
16             s^(-1)  7.0         15.0  
...               ...  ...          ...  
333508       

### Explore and reformat data entries

#### `EntryID`

In [37]:
len(np.unique(df['EntryID']))

28056

Only 28k out of all 74k EntryIDs in the database have an associated $k_{cat}$ value.

#### `PubMedID`

These are parsed as floats but should be integers

In [38]:
np.unique(df['PubMedID'].astype(str))

array(['10021417.0', '10026144.0', '10026151.0', ..., '9989239.0',
       '9990323.0', 'nan'], dtype=object)

In [39]:
df['PubMedID']

3         12962477.0
4         15311930.0
7         15740745.0
13        15740745.0
16        15740745.0
             ...    
333508    12912903.0
333511    12912903.0
333512    12912903.0
333976    11141073.0
333983    11141073.0
Name: PubMedID, Length: 28220, dtype: float64

In [40]:
df['PubMedID'] = df['PubMedID'].astype('Int64')

#### `Substrate`

In the end we reduce the dataset to $k_{cat}$ values only. If I understand correctly, in the DLKcat paper, the authors associate the $k_{cat}$ value with the substrate that has a measured $K_M$ value. If the $K_M$ values are measured for multiple substrates in the same reaction, then this would result in separate data inputs of the form $\text{substrate}_1 \rightarrow k_{cat}$ and $\text{substrate}_2 \rightarrow k_{cat}$ (in the neural net framework this can lead to different `kcat` predictions depending on the input substrate). We want to construct an input vector that includes all the substrates involved in the reaction (ignoring whether it has a measured $K_M$ value) in the form $[\text{substrate}_1, \text{substrate}_2, \dotsc] \rightarrow k_{cat}$. 

This is mentioned in the paper's discussion: "Another challenge relates to reactions involving multiple substrates and those catalysed by heteromeric enzyme complexes. The multiple substrate SMILES and protein sequences that can be defined for such reactions can all function with DLKcat, thereby yielding multiple predicted kcat values for one reaction. We currently select the maximum kcat values in those cases, but it would be favourable to devise an approach that can predict one kcat value for each multi-substrate and/or heteromeric enzyme."

Each entry is a list of substrates separated by `;` -- split into an array of individual substrate strings.

In [41]:
df['Substrate'] = df['Substrate'].apply(lambda x: np.sort(x.split(';')))

#### `EnzymeType`

In [42]:
df['EnzymeType']

3         wildtype variant DSAI (N76D/N87S/S103A/V104I)
4                                       mutant MDH G81S
7                                         wildtype DHFR
13                                     mutant DHFR W30A
16                                     mutant DHFR W30E
                              ...                      
333508                                         wildtype
333511                                     mutant H225Q
333512                                     mutant Y259F
333976                                    wildtype P26h
333983                                    wildtype P26h
Name: EnzymeType, Length: 28220, dtype: object

In [43]:
mutant_namelist = df['EnzymeType'][df['EnzymeType'].astype(str).apply(lambda x: 'mutant' in x)]

In [ ]:
# code from DLKcat
import re

for name in df['EnzymeType'].astype(str): 
    if 'wildtype' in name:
        print (name +' -> wildtype')
    else :
    # if 'mutant' in EnzymeType or 'mutated' in EnzymeType:
        mutant = re.findall('[A-Z]\d+[A-Z]', name)  # re is of great use
        print(name + ' -> '+ '/'.join(mutant))

The modifications given above that simplify the mutant/wild-type descriptors are performed in the DLKcat scripts. Omitting it now to keep things general.

#### `Organism`

In [45]:
np.unique(df['Organism'])

array(['Abax parallelepipedus', 'Acanthamoeba polyphaga mimivirus',
       'Acetobacter aceti', 'Acetobacter xylinus',
       'Achromobacter cycloclastes', 'Achromobacter xylosoxidans',
       'Acidianus sp.', 'Acidithiobacillus thiooxidans',
       'Acidobacteriaceae', 'Acidovorax citrulli',
       'Acinetobacter baumannii', 'Acinetobacter baylyi',
       'Acinetobacter calcoaceticus',
       'Acinetobacter calcoaceticus (subsp. anitratus)',
       'Acremonium persicinum', 'Actinidia chinensis',
       'Actinidia eriantha', 'Actinobacillus succinogenes',
       'Actinomadura sp.', 'Actinoplanes missouriensis', 'Aedes aegypti',
       'Aeromonas hydrophila', 'Aeromonas sobria', 'Aeropyrum pernix',
       'Agaricus bisporus', 'Agarivorans albus', 'Agkistrodon bilineatus',
       'Agkistrodon rhodostoma', 'Agrobacterium tumefaciens',
       'Agrobacterium vitis', 'Agrocybe aegerita',
       'Akkermansia muciniphila', 'Alcaligenes faecalis',
       'Alicyclobacillus acidocaldarius',
     

Keeping the organism list as is.

#### `UniprotID`

In [46]:
np.unique(df['UniprotID'][df['UniprotID'].apply(lambda x: type(x) != str)].astype(str))

array(['nan'], dtype=object)

Some entries do not have an associated UniprotID.

In [47]:
df.loc[[3]]

EntryID  PubMedID         Organism       Substrate  \
3        1  12962477  Bacillus lentus  [H2O, Peptide]   

                                      EnzymeType  Enzymename UniprotID  \
3  wildtype variant DSAI (N76D/N87S/S103A/V104I)  subtilisin       NaN   

    ECNumber  parameter.startValue  parameter.endValue  \
3  3.4.21.62                 220.0                 NaN   

   parameter.standardDeviation parameter.unit   pH  Temperature  
3                          NaN         s^(-1)  7.5         25.0

Some entries have multiple associated UniprotIDs (represent heteromeric protein complexes?). Detect such entries and list the UniprotIDs as an array

In [49]:
df['UniprotID'] = df['UniprotID'].apply(lambda x: np.empty(0) if pd.isnull(x) else np.sort(x.split(' ')))

In [50]:
df[df['UniprotID'].astype(str).apply(lambda x : len(x)) == 2]

EntryID  PubMedID                     Organism  \
3             1  12962477              Bacillus lentus   
394       10048  14617643         Pseudomonas reinekei   
397       10049  14617643         Pseudomonas reinekei   
528       10082  11284686              Giardia lamblia   
532       10083  11284686              Giardia lamblia   
...         ...       ...                          ...   
332145      990   9890907                   Sus scrofa   
333078     9793  15610039  Yersinia pseudotuberculosis   
333508     9874  12912903     Arthrobacter globiformis   
333511     9875  12912903     Arthrobacter globiformis   
333512     9876  12912903     Arthrobacter globiformis   

                                                Substrate  \
3                                          [H2O, Peptide]   
394       [H2O, trans-4-Carboxymethylenebut-2-en-4-olide]   
397                          [4-Fluoromuconolactone, H2O]   
528     [5-Phospho-alpha-D-ribose 1-diphosphate, Guanine]   
532     [5-Phospho-alpha-D-ribose 1-diphosphate, Guanine]   
...                                                   ...   
332145            [Dehydroascorbate, Reduced glutathione]   
333078            [GDP-4-dehydro-D-rhamnose, L-Glutamate]   
333508                     [H2O, N,N-Dimethylglycine, O2]   
333511                     [H2O, N,N-Dimethylglycine, O2]   
333512                     [H2O, N,N-Dimethylglycine, O2]   

                                           EnzymeType  \
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)   
394                                          wildtype   
397                                          wildtype   
528                                          wildtype   
532                                       mutant S96A   
...                                               ...   
332145                                       wildtype   
333078                                  wildtype ColD   
333508                                       wildtype   
333511                                   mutant H225Q   
333512                                   mutant Y259F   

                                   Enzymename UniprotID   ECNumber  \
3                                  subtilisin        []  3.4.21.62   
394              carboxymethylenebutenolidase        []   3.1.1.45   
397              carboxymethylenebutenolidase        []   3.1.1.45   
528                      Pentosyltransferases        []      2.4.2   
532                      Pentosyltransferases        []      2.4.2   
...                                       ...       ...        ...   
332145  glutathione dehydrogenase (ascorbate)        []    1.8.5.1   
333078                                    NaN        []        NaN   
333508                dimethylglycine oxidase        []   1.5.3.10   
333511                dimethylglycine oxidase        []   1.5.3.10   
333512                dimethylglycine oxidase        []   1.5.3.10   

        parameter.startValue  parameter.endValue  parameter.standardDeviation  \
3                 220.000000                 NaN                          NaN   
394              2650.000000                 NaN                   141.666700   
397              1633.333000                 NaN                    36.666670   
528                76.700000                 NaN                     2.500000   
532                55.700000                 NaN                     4.200000   
...                      ...                 ...                          ...   
332145              7.566667                 NaN                     0.283333   
333078              0.600000                 NaN                          NaN   
333508             10.600000                 NaN                          NaN   
333511              4.350000                 NaN                          NaN   
333512              0.004700                 NaN                          NaN   

       parameter.unit   pH  Temperature  
3              s^(-1)  7.5         25

#### `ECNumber`

Multiple ECNumbers may be associated with an entry -- putting them in arrays.

In [51]:
df['ECNumber'] = df['ECNumber'].apply(lambda x: np.empty(0) if pd.isnull(x) else np.sort(x.split(' ')))

In [52]:
mask = (df['ECNumber'].apply(lambda x: len(x) == 0)) & (df['UniprotID'].apply(lambda x: len(x) == 0))

In [53]:
np.sum(mask)

np.int64(20)

Remove entries that are missing both `ECNumber` and `UniprotID` (have no associated enzyme information) 

In [54]:
df = df.loc[~mask]

#### `parameter`

Some database entries have multiple associated $k_{cat}$ values. In DLKcat, the duplicates are removed by selecting the maximum $k_{cat}$ value as the ground truth. This is not necessarily the best approach as the multiple values may reflect experimental uncertainty or alternative specific experimental conditions. However, this is difficult to parse in an automated way, and dealing with such values more rigorously may require manual labour...

Finally, note that in some cases the $k_{cat}$ values have not only an associated `parameter.startValue` but also an extra `parameter.endValue` (indicating an interval of possible values) or an associated `parameter.standardDeviation`. It's questionable whether these could be used as extra inputs.

In [55]:
i = 0
for entryID in np.unique(df['EntryID']):
    n = np.sum(df['EntryID'] == entryID)
    if n > 1:
        i += 1
        print('EntryID %d has %d kcat values' % (entryID, n))
print('Number of entries with more than one kcat: %d' %i)

EntryID 294 has 2 kcat values
EntryID 295 has 2 kcat values
EntryID 296 has 2 kcat values
EntryID 297 has 2 kcat values
EntryID 428 has 3 kcat values
EntryID 443 has 2 kcat values
EntryID 444 has 2 kcat values
EntryID 445 has 2 kcat values
EntryID 486 has 2 kcat values
EntryID 985 has 2 kcat values
EntryID 986 has 2 kcat values
EntryID 1075 has 2 kcat values
EntryID 1868 has 2 kcat values
EntryID 1869 has 2 kcat values
EntryID 1870 has 2 kcat values
EntryID 1871 has 2 kcat values
EntryID 1973 has 2 kcat values
EntryID 1974 has 2 kcat values
EntryID 1975 has 2 kcat values
EntryID 1976 has 2 kcat values
EntryID 1977 has 2 kcat values
EntryID 1978 has 2 kcat values
EntryID 1979 has 2 kcat values
EntryID 1980 has 2 kcat values
EntryID 1981 has 2 kcat values
EntryID 1982 has 2 kcat values
EntryID 1983 has 2 kcat values
EntryID 1984 has 2 kcat values
EntryID 2872 has 2 kcat values
EntryID 2873 has 2 kcat values
EntryID 2874 has 2 kcat values
EntryID 2875 has 2 kcat values
EntryID 2876 has 2 

A few are due to duplicated data entries. Remove those

In [56]:
df = df.loc[df.astype(str).drop_duplicates().index]

In [57]:
len(df)

28195

In [58]:
i = 0
for entryID in np.unique(df['EntryID']):
    n = np.sum(df['EntryID'] == entryID)
    if n > 1:
        i += 1
        print('EntryID %d has %d kcat values' % (entryID, n))
print('Number of entries with more than one kcat: %d' %i)

EntryID 294 has 2 kcat values
EntryID 295 has 2 kcat values
EntryID 296 has 2 kcat values
EntryID 297 has 2 kcat values
EntryID 428 has 3 kcat values
EntryID 443 has 2 kcat values
EntryID 444 has 2 kcat values
EntryID 445 has 2 kcat values
EntryID 486 has 2 kcat values
EntryID 985 has 2 kcat values
EntryID 986 has 2 kcat values
EntryID 1075 has 2 kcat values
EntryID 1868 has 2 kcat values
EntryID 1869 has 2 kcat values
EntryID 1870 has 2 kcat values
EntryID 1871 has 2 kcat values
EntryID 1973 has 2 kcat values
EntryID 1974 has 2 kcat values
EntryID 1975 has 2 kcat values
EntryID 1976 has 2 kcat values
EntryID 1977 has 2 kcat values
EntryID 1978 has 2 kcat values
EntryID 1979 has 2 kcat values
EntryID 1980 has 2 kcat values
EntryID 1981 has 2 kcat values
EntryID 1982 has 2 kcat values
EntryID 1983 has 2 kcat values
EntryID 1984 has 2 kcat values
EntryID 2873 has 2 kcat values
EntryID 2874 has 2 kcat values
EntryID 2875 has 2 kcat values
EntryID 2876 has 2 kcat values
EntryID 2878 has 2 

#### `parameter.endValue`

List the number of $k_{cat}$ values with an associated `endValue`

In [59]:
np.sum(~df['parameter.endValue'].isnull())

np.int64(103)

Very few entries exist, so ignoring `endValue` seems to be the way to go. Remove the column completely

In [60]:
df = df.drop('parameter.endValue', axis=1)

#### `parameter.standardDeviation`

List the number of $k_{cat}$ values with an associated `standardDeviation`

In [61]:
np.sum(~df['parameter.standardDeviation'].isnull())

np.int64(16020)

Standard deviation is given for a relatively good chunk -- might be worth considering.

#### `parameter.unit`

Units are not given in certain cases.

In [62]:
df[df['parameter.unit'].apply(lambda x : pd.isnull(x))]

EntryID  PubMedID                Organism  \
149829    41967   9288931   Flaveria chloraefolia   
149838    41969   9288931   Flaveria chloraefolia   
149849    41971   9288931   Flaveria chloraefolia   
149861    41973   9288931   Flaveria chloraefolia   
149877    41975   9288931   Flaveria chloraefolia   
149887    41977   9288931   Flaveria chloraefolia   
149897    41979   9288931   Flaveria chloraefolia   
149910    41981   9288931   Flaveria chloraefolia   
149921    41983   9288931   Flaveria chloraefolia   
149935    41985   9288931   Flaveria chloraefolia   
204795    53222   1974463  Bacillus licheniformis   
303617    73389  18237196        SARS coronavirus   

                                      Substrate  \
149829  [3'-Phosphoadenylyl sulfate, Quercetin]   
149838  [3'-Phosphoadenylyl sulfate, Quercetin]   
149849  [3'-Phosphoadenylyl sulfate, Quercetin]   
149861  [3'-Phosphoadenylyl sulfate, Quercetin]   
149877  [3'-Phosphoadenylyl sulfate, Quercetin]   
149887  [3'-Phosphoadenylyl sulfate, Quercetin]   
149897  [3'-Phosphoadenylyl sulfate, Quercetin]   
149910  [3'-Phosphoadenylyl sulfate, Quercetin]   
149921  [3'-Phosphoadenylyl sulfate, Quercetin]   
149935  [3'-Phosphoadenylyl sulfate, Quercetin]   
204795                      [H2O, Penicillin V]   
303617       [Dabcyl-KTSAVLQSGFRKME-Edans, H2O]   

                            EnzymeType                        Enzymename  \
149829                     mutant P120       flavonol 3-sulfotransferase   
149838                     mutant L120       flavonol 3-sulfotransferase   
149849                     mutant K126       flavonol 3-sulfotransferase   
149861                     mutant R134       flavonol 3-sulfotransferase   
149877                     mutant T134       flavonol 3-sulfotransferase   
149887                     mutant G134       flavonol 3-sulfotransferase   
149897                     mutant A120       flavonol 3-sulfotransferase   
149910                     mutant H175       flavonol 3-sulfotransferase   
149921                     mutant Q186       flavonol 3-sulfotransferase   
149935                     mutant A186       flavonol 3-sulfotransferase   
204795                    mutant K234E                    beta-lactamase   
303617  wildtype 3CLpro (aa 3241-3546)  SARS coronavirus main proteinase   

                               UniprotID     ECNumber  parameter.startValue  \
149829                          [P52836]   [2.8.2.25]                  0.53   
149838                          [P52836]   [2.8.2.25]                  0.69   
149849                          [P52836]   [2.8.2.25]                  0.92   
149861                          [P52836]   [2.8.2.25]                  1.66   
149877                          [P52836]   [2.8.2.25]                  0.53   
149887                          [P52836]   [2.8.2.25]                  0.27   
149897                          [P52836]   [2.8.2.25]                  0.41   
149910                          [P52836]   [2.8.2.25]                  1.36   
149921                          [P52836]   [2.8.2.25]                  0.26   
149935                          [P52836]   [2.8.2.25]                  0.06   
204795                          [P00808]    [3.5.2.6]                 37.30   
303617  [P0C6U8, P0C6U8, P0C6X7, P0C6X7]  [3.4.22.69]                  3.70   

        parameter.standardDeviation parameter.unit   pH  Temperature  
149829                          NaN            NaN  7.5         25.0  
149838                          NaN            NaN  7.5         25.0  
149849                          NaN            NaN  7.5         25.0  
149861                          NaN            NaN  7.5         25.0  
149877                          NaN            NaN  7.5         25.0  
149887                          NaN            NaN  7.5         25.0  
149897                          NaN            NaN  7.5         25.0  
149910                          NaN            NaN  7.5         25.0 

Having checked the paper for EntryID: 41969, the units are given as $s^{-1}$, so it's a human error in this case. Probably okay to assume that all `NaN` units can be converted to $s^{-1}$ (as done in DLKcat). 

Some $k_{cat}$ values have weird units associated with them. Some appear to be an error, also leading to wrong values due to an automatic unit conversion step.

In [63]:
np.unique(df['parameter.unit'].astype(str))

array(['J/mol', 'M', 'M^(-1)*s^(-1)', 'katal', 'mol*s^(-1)*g^(-1)',
       'mol*s^(-1)*mol^(-1)', 'nan', 's^(-1)', 's^(-1)*g^(-1)'],
      dtype=object)

In [64]:
df[df['parameter.unit'] == 'J/mol']

EntryID  PubMedID        Organism         Substrate  \
270394    66706  17335913  Vaccinia virus  [ATP, Thymidine]   

                 EnzymeType        Enzymename UniprotID    ECNumber  \
270394  wildtype His-tagged  thymidine kinase  [P68563]  [2.7.1.21]   

        parameter.startValue  parameter.standardDeviation parameter.unit   pH  \
270394              0.921096                     0.167472          J/mol  8.0   

        Temperature  
270394         37.0

Typo having checked the paper (should be $s^{-1}$)

In [65]:
df[df['parameter.unit'] == 'M']

EntryID  PubMedID                 Organism  \
110270    33792   8175784        Rattus norvegicus   
110292    33796   8175784        Rattus norvegicus   
110313    33800   8175784        Rattus norvegicus   
110330    33804   8175784        Rattus norvegicus   
200401    52323  14640970        Bacillus subtilis   
200408    52324  14640970        Bacillus subtilis   
254555    63591  17848100  Mycobacterium smegmatis   

                                          Substrate  \
110270     [5alpha-Androstane-3,17-dione, H+, NADH]   
110292     [5alpha-Androstane-3,17-dione, H+, NADH]   
110313     [5alpha-Androstane-3,17-dione, H+, NADH]   
110330     [5alpha-Androstane-3,17-dione, H+, NADH]   
200401        [Glp-Ala-Ala-Leu-p-Nitroanilide, H2O]   
200408        [Glp-Ala-Ala-Leu-p-Nitroanilide, H2O]   
254555  [Enzyme-cysadenylate-PPi complex, GlcN-Ins]   

                                 EnzymeType  \
110270    mutant isoenzyme 3alpha-HSD C170A   
110292    mutant isoenzyme 3alpha-HSD C217A   
110313    mutant isoenzyme 3alpha-HSD C242A   
110330    mutant isoenzyme 3alpha-HSD Y205F   
200401                             wildtype   
200408  wildtype poly(acrylic acid) complex   
254555             wildtype MshC His-tagged   

                                               Enzymename UniprotID  \
110270   3alpha-hydroxysteroid dehydrogenase (A-specific)  [P23457]   
110292   3alpha-hydroxysteroid dehydrogenase (A-specific)  [P23457]   
110313   3alpha-hydroxysteroid dehydrogenase (A-specific)  [P23457]   
110330   3alpha-hydroxysteroid dehydrogenase (A-specific)  [P23457]   
200401                                         subtilisin        []   
200408                                         subtilisin        []   
254555  L-cysteine:1D-myo-inositol 2-amino-2-deoxy-alp...        []   

           ECNumber  parameter.startValue  parameter.standardDeviation  \
110270  [1.1.1.213]              0.000025                 7.000000e-07   
110292  [1.1.1.213]              0.000026                 1.100000e-06   
110313  [1.1.1.213]              0.000022                 4.000000e-07   
110330  [1.1.1.213]              0.000022                 7.000000e-07   
200401  [3.4.21.62]              0.069300                 3.500000e-03   
200408  [3.4.21.62]              0.051000                 2.600000e-03   
254555   [6.3.1.13]              0.005200                 8.000000e-04   

       parameter.unit   pH  Temperature  
110270              M  7.0         25.0  
110292              M  7.0         25.0  
110313              M  7.0         25.0  
110330              M  7.0         25.0  
200401              M  8.2         37.0  
200408              M  8.2         37.0  
254555              M  7.8         25.0

Clean up the units. Keep only `s^(-1)` and convert `mol*s^(-1)*mol^(-1)` and `NaN` to `s^(-1)`

In [66]:
mask = ((df['parameter.unit'].isnull()) | (df['parameter.unit'] == 'mol*s^(-1)*mol^(-1)'))
df.loc[mask, 'parameter.unit'] = 's^(-1)'

In [67]:
mask = df['parameter.unit'] == 's^(-1)'
df = df.loc[mask]

#### `pH`

In [68]:
np.sum(~df['pH'].isnull())

np.int64(27334)

#### `Temperature`

In [69]:
np.sum(~df['Temperature'].isnull())

np.int64(25953)

Most entries seem to have an associated pH and temperature.

# Save dataset

In [70]:
df

EntryID  PubMedID                  Organism  \
3             1  12962477           Bacillus lentus   
4            10  15311930        Pseudomonas putida   
7           100  15740745          Escherichia coli   
13          101  15740745          Escherichia coli   
16          102  15740745          Escherichia coli   
...         ...       ...                       ...   
333508     9874  12912903  Arthrobacter globiformis   
333511     9875  12912903  Arthrobacter globiformis   
333512     9876  12912903  Arthrobacter globiformis   
333976     9970  11141073      Mesocricetus auratus   
333983     9971  11141073      Mesocricetus auratus   

                                            Substrate  \
3                                      [H2O, Peptide]   
4             [(S)-Mandelate, Riboflavin-5-phosphate]   
7                          [Dihydrofolate, H+, NADPH]   
13                         [Dihydrofolate, H+, NADPH]   
16                         [Dihydrofolate, H+, NADPH]   
...                                               ...   
333508                 [H2O, N,N-Dimethylglycine, O2]   
333511                 [H2O, N,N-Dimethylglycine, O2]   
333512                 [H2O, N,N-Dimethylglycine, O2]   
333976  [5alpha-Androstane-17beta-ol-3-one, H+, NADH]   
333983   [5alpha-Androstane-3alpha,17beta-diol, NAD+]   

                                           EnzymeType  \
3       wildtype variant DSAI (N76D/N87S/S103A/V104I)   
4                                     mutant MDH G81S   
7                                       wildtype DHFR   
13                                   mutant DHFR W30A   
16                                   mutant DHFR W30E   
...                                               ...   
333508                                       wildtype   
333511                                   mutant H225Q   
333512                                   mutant Y259F   
333976                                  wildtype P26h   
333983                                  wildtype P26h   

                         Enzymename UniprotID     ECNumber  \
3                        subtilisin        []  [3.4.21.62]   
4       (S)-mandelate dehydrogenase  [P20932]  [1.1.99.31]   
7           dihydrofolate reductase  [P0ABQ4]    [1.5.1.3]   
13          dihydrofolate reductase  [P0ABQ4]    [1.5.1.3]   
16          dihydrofolate reductase  [P0ABQ4]    [1.5.1.3]   
...                             ...       ...          ...   
333508      dimethylglycine oxidase        []   [1.5.3.10]   
333511      dimethylglycine oxidase        []   [1.5.3.10]   
333512      dimethylglycine oxidase        []   [1.5.3.10]   
333976   carbonyl reductase (NADPH)  [Q9Z245]  [1.1.1.184]   
333983   carbonyl reductase (NADPH)  [Q9Z245]  [1.1.1.184]   

        parameter.startValue  parameter.standardDeviation parameter.unit   pH  \
3                   220.0000                          NaN         s^(-1)  7.5   
4                     2.8000                          0.1         s^(-1)  7.5   
7                     6.1800                          NaN         s^(-1)  7.0   
13                   14.5000                          NaN         s^(-1)  7.0   
16                   11.5800                          NaN         s^(-1)  7.0   
...                      ...                          ...            ...  ...   
333508               10.6000                          NaN         s^(-1)  8.5   
333511                4.3500                          NaN         s^(-1)  8.5   
333512                0.0047                          NaN         s^(-1)  8.5   
333976                0.8500                          NaN         s^(-1)  7.0   
333983                1.2800                          NaN         s^(-1)  7.0   

        Temperature  
3              25.0  
4              20.0  
7              15.0  
13             15.0  
16             15.0  
...             ...  
333508         25.0  
333511         25.0  
333512         25.0  
333976         25.0  
333983         25

In [71]:
df = df.rename(columns={'parameter.startValue': 'Value'})
df = df.rename(columns={'parameter.standardDeviation': 'StandardDeviation'})
df = df.rename(columns={'UniprotID': 'UniProtID'})
df = df.rename(columns={'Enzymename': 'EnzymeName'})
df = df.drop('parameter.unit', axis=1)

In [75]:
df = df[['PubMedID', 'Substrate', 'ECNumber', 'EnzymeName', 'EnzymeType', 'UniProtID', 'pH', 'Temperature', 'Value', 'StandardDeviation']]

In [76]:
df

PubMedID                                      Substrate     ECNumber  \
0      12962477                                 [H2O, Peptide]  [3.4.21.62]   
1      15311930        [(S)-Mandelate, Riboflavin-5-phosphate]  [1.1.99.31]   
2      15740745                     [Dihydrofolate, H+, NADPH]    [1.5.1.3]   
3      15740745                     [Dihydrofolate, H+, NADPH]    [1.5.1.3]   
4      15740745                     [Dihydrofolate, H+, NADPH]    [1.5.1.3]   
...         ...                                            ...          ...   
28124  12912903                 [H2O, N,N-Dimethylglycine, O2]   [1.5.3.10]   
28125  12912903                 [H2O, N,N-Dimethylglycine, O2]   [1.5.3.10]   
28126  12912903                 [H2O, N,N-Dimethylglycine, O2]   [1.5.3.10]   
28127  11141073  [5alpha-Androstane-17beta-ol-3-one, H+, NADH]  [1.1.1.184]   
28128  11141073   [5alpha-Androstane-3alpha,17beta-diol, NAD+]  [1.1.1.184]   

                        EnzymeName  \
0                       subtilisin   
1      (S)-mandelate dehydrogenase   
2          dihydrofolate reductase   
3          dihydrofolate reductase   
4          dihydrofolate reductase   
...                            ...   
28124      dimethylglycine oxidase   
28125      dimethylglycine oxidase   
28126      dimethylglycine oxidase   
28127   carbonyl reductase (NADPH)   
28128   carbonyl reductase (NADPH)   

                                          EnzymeType UniProtID   pH  \
0      wildtype variant DSAI (N76D/N87S/S103A/V104I)        []  7.5   
1                                    mutant MDH G81S  [P20932]  7.5   
2                                      wildtype DHFR  [P0ABQ4]  7.0   
3                                   mutant DHFR W30A  [P0ABQ4]  7.0   
4                                   mutant DHFR W30E  [P0ABQ4]  7.0   
...                                              ...       ...  ...   
28124                                       wildtype        []  8.5   
28125                                   mutant H225Q        []  8.5   
28126                                   mutant Y259F        []  8.5   
28127                                  wildtype P26h  [Q9Z245]  7.0   
28128                                  wildtype P26h  [Q9Z245]  7.0   

       Temperature     Value  StandardDeviation  
0             25.0  220.0000                NaN  
1             20.0    2.8000                0.1  
2             15.0    6.1800                NaN  
3             15.0   14.5000                NaN  
4             15.0   11.5800                NaN  
...            ...       ...                ...  
28124         25.0   10.6000                NaN  
28125         25.0    4.3500                NaN  
28126         25.0    0.0047                NaN  
28127         25.0    0.8500                NaN  
28128         25.0    1.2800                NaN  

[28129 rows x 10 columns]

In [84]:
fdir = '../../../data/databases/Sabio-RK/'

In [85]:
df = df.reset_index(drop=True)

In [86]:
df.to_csv(fdir+'kcats.csv', index=False)

In [87]:
df.to_json(fdir+'kcats.json', index=False)